### ISSUE TITLE
RTDETR

### ISSUE THREAD

**Issue by MarcoPrassel** [marcoslucianops/DeepStream-Yolo#476]
Hi @marcoslucianops , 
I am trying to integrate Ultralytics' custom RT-DETR model with Deepstream, but I am having problems with the layer parser from ONNX. Could you please help me ?
This is my ONNX's output
![image](https://github.com/marcoslucianops/DeepStream-Yolo/assets/109653912/46c3b002-c082-4d80-8163-5d6b2736da55)

**Comment by rpatidar** [comment]
Did it work for you ?

**Comment by MarcoPrassel** [comment]
@rpatidar  No, it did not work. I also tried to take advantage of the Nvidia Deepstream solutions, but the output and input layers have to be changed. I hope @marcoslucianops  can help me.

**Comment by rpatidar** [comment]
Ohk , I am able to evaluate the model with tensort model  https://github.com/lyuwenyu/RT-DETR/blob/main/benchmark/trtinfer.py example shared it  here. 

thought existing default plugin would have worked but with the deepstream, looks like it didn't  was giving up some model and parameter mismatched error. 

looking for your expertise  @marcoslucianops  on how can we use the model with the deepstream

**Comment by marcoslucianops** [comment]
I will add the support today. For now, only for this repo: https://github.com/lyuwenyu/RT-DETR

**Comment by marcoslucianops** [comment]
https://github.com/marcoslucianops/DeepStream-Yolo/blob/master/docs/RTDETR.md

**Comment by rpatidar** [comment]
I can confirms, I am able to use it, haven't tested all batching feature and other stuffs but for single stream its working as expected. 

Thanks you  so much.  I was almost lost on how to get it working.

**Comment by tms2003** [comment]
@marcoslucianops  I noticed that there is no benchmark yet. Does it perform well?

**Comment by rpatidar** [comment]
I was able to execute the ultralytics  based model by adding following batch to `export_yolov8.py` 

``` class DeepStreamOutput(nn.Module):
-    def __init__(self):
+    def __init__(self, img_size):
+        self.img_size = img_size
         super().__init__()

     def forward(self, x):
-        x = x.transpose(1, 2)
+        #x = x.transpose(1, 2)
         boxes = x[:, :, :4]
+        boxes[:, :, [0, 2]] *= self.img_size[1]
+        boxes[:, :, [1, 3]] *= self.img_size[0]
         scores, classes = torch.max(x[:, :, 4:], 2, keepdim=True)
         classes = classes.float()
         return boxes, scores, classes
@@ -64,7 +67,7 @@ def main(args):
             f.write(name + '\n')
         f.close()

-    model = nn.Sequential(model, DeepStreamOutput())
+    model = nn.Sequential(model, DeepStreamOutput((640,640)))
```

**Comment by marcoslucianops** [comment]
Thank you @rpatidar, the multiply to the img_size did the trick.

**Comment by citywatchnz** [comment]
Hi, 

Has the RT-DETR been tested on the Jetson NX? When I try the a trained onnx model on the Jetson NX, the conversion failed with the below message. It works fine on the PC - converted and running. I am using Deepstream 6.0.

> WARNING: [TRT]: onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
WARNING: [TRT]: onnx2trt_utils.cpp:392: One or more weights outside the range of INT32 was clamped
ERROR: [TRT]: ModelImporter.cpp:773: While parsing node number 444 [GridSample -> "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"]:
ERROR: [TRT]: ModelImporter.cpp:774: --- Begin node ---
ERROR: [TRT]: ModelImporter.cpp:775: input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_5_output_0"
input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_6_output_0"
output: "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"
name: "/0/decoder/decoder/layers.0/cross_attn/GridSample"
op_type: "GridSample"
attribute {
  name: "align_corners"
  i: 0
  type: INT
}
attribute {
  name: "mode"
  s: "bilinear"
  type: STRING
}
attribute {
  name: "padding_mode"
  s: "zeros"
  type: STRING
}
ERROR: [TRT]: ModelImporter.cpp:776: --- End node ---
ERROR: [TRT]: ModelImporter.cpp:779: ERROR: builtin_op_importers.cpp:4870 In function importFallbackPluginImporter:
[8] Assertion failed: creator && "Plugin not found, are the plugin name, version, and namespace correct?"
ERROR: Failed to parse onnx file
ERROR: failed to build network since parsing model errors.

**Comment by IronmanVsThanos** [comment]
> Hi,
> 
> Has the RT-DETR been tested on the Jetson NX? When I try the a trained onnx model on the Jetson NX, the conversion failed with the below message. It works fine on the PC - converted and running. I am using Deepstream 6.0.
> 
> > WARNING: [TRT]: onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
> > WARNING: [TRT]: onnx2trt_utils.cpp:392: One or more weights outside the range of INT32 was clamped
> > ERROR: [TRT]: ModelImporter.cpp:773: While parsing node number 444 [GridSample -> "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"]:
> > ERROR: [TRT]: ModelImporter.cpp:774: --- Begin node ---
> > ERROR: [TRT]: ModelImporter.cpp:775: input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_5_output_0"
> > input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_6_output_0"
> > output: "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"
> > name: "/0/decoder/decoder/layers.0/cross_attn/GridSample"
> > op_type: "GridSample"
> > attribute {
> > name: "align_corners"
> > i: 0
> > type: INT
> > }
> > attribute {
> > name: "mode"
> > s: "bilinear"
> > type: STRING
> > }
> > attribute {
> > name: "padding_mode"
> > s: "zeros"
> > type: STRING
> > }
> > ERROR: [TRT]: ModelImporter.cpp:776: --- End node ---
> > ERROR: [TRT]: ModelImporter.cpp:779: ERROR: builtin_op_importers.cpp:4870 In function importFallbackPluginImporter:
> > [8] Assertion failed: creator && "Plugin not found, are the plugin name, version, and namespace correct?"
> > ERROR: Failed to parse onnx file
> > ERROR: failed to build network since parsing model errors.

I have samae question with you

**Comment by IronmanVsThanos** [comment]
> Hi,
> 
> Has the RT-DETR been tested on the Jetson NX? When I try the a trained onnx model on the Jetson NX, the conversion failed with the below message. It works fine on the PC - converted and running. I am using Deepstream 6.0.
> 
> > WARNING: [TRT]: onnx2trt_utils.cpp:366: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
> > WARNING: [TRT]: onnx2trt_utils.cpp:392: One or more weights outside the range of INT32 was clamped
> > ERROR: [TRT]: ModelImporter.cpp:773: While parsing node number 444 [GridSample -> "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"]:
> > ERROR: [TRT]: ModelImporter.cpp:774: --- Begin node ---
> > ERROR: [TRT]: ModelImporter.cpp:775: input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_5_output_0"
> > input: "/0/decoder/decoder/layers.0/cross_attn/Reshape_6_output_0"
> > output: "/0/decoder/decoder/layers.0/cross_attn/GridSample_output_0"
> > name: "/0/decoder/decoder/layers.0/cross_attn/GridSample"
> > op_type: "GridSample"
> > attribute {
> > name: "align_corners"
> > i: 0
> > type: INT
> > }
> > attribute {
> > name: "mode"
> > s: "bilinear"
> > type: STRING
> > }
> > attribute {
> > name: "padding_mode"
> > s: "zeros"
> > type: STRING
> > }
> > ERROR: [TRT]: ModelImporter.cpp:776: --- End node ---
> > ERROR: [TRT]: ModelImporter.cpp:779: ERROR: builtin_op_importers.cpp:4870 In function importFallbackPluginImporter:
> > [8] Assertion failed: creator && "Plugin not found, are the plugin name, version, and namespace correct?"
> > ERROR: Failed to parse onnx file
> > ERROR: failed to build network since parsing model errors.

have you slove this problem?

**Comment by IronmanVsThanos** [comment]
> https://github.com/marcoslucianops/DeepStream-Yolo/blob/master/docs/RTDETR.md

404 - page not found

**Comment by prashant-dn** [comment]
> > https://github.com/marcoslucianops/DeepStream-Yolo/blob/master/docs/RTDETR.md
> 
> 404 - page not found
![Screenshot 2024-08-14 at 11 30 49 AM](https://github.com/user-attachments/assets/f395ca00-ebf7-4484-86ef-7867d2dbd4f1)

there are 3 different READMEs now.

**Comment by marcoslucianops** [comment]
@citywatchnz @IronmanVsThanos, try with lower `opset` on the exporter file.
